From 71abfab0ba93982b5dcc5effc61b7343b2a653e4 Mon Sep 17 00:00:00 2001
From: hariprasad <hkelam@marvell.com>
Date: Thu, 11 Jul 2019 15:51:41 +0530
Subject: [PATCH 0233/1921] octeontx2-af: Delete range of mcam entries

Extend npc_delete_flow functionality to delete range of mcam
entries.This feature will be used to delete rules associated with
specific filters (ntuple,unicast etc).

Change-Id: Ia8e8713c5ccc77f5abe654a1c7c19eb342affe65
Signed-off-by: hariprasad <hkelam@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/12560
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/mbox.h   |  2 ++
 .../net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c | 14 ++++++++++++++
 2 files changed, 16 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/mbox.h b/drivers/net/ethernet/marvell/octeontx2/af/mbox.h
index 0685123994c5..926cc09fca23 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/mbox.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/mbox.h
@@ -1306,6 +1306,8 @@ struct npc_install_flow_rsp {
 struct npc_delete_flow_req {
 	struct mbox_msghdr hdr;
 	u16 entry;
+	u16 start;/*Disable range of entries */
+	u16 end;
 	u8 all; /* PF + VFs */
 };
 
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c
index c331aad26e7f..eaf39b4da14a 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c
@@ -961,6 +961,20 @@ int rvu_mbox_handler_npc_delete_flow(struct rvu *rvu,
 	u16 pcifunc = req->hdr.pcifunc;
 	int err;
 
+	if (req->end) {
+		list_for_each_entry_safe(iter, tmp, &mcam->mcam_rules, list) {
+			if (iter->owner == pcifunc &&
+			    iter->entry >= req->start &&
+			    iter->entry <= req->end) {
+				err = npc_delete_flow(rvu, iter->entry,
+						      pcifunc);
+				if (err)
+					return err;
+			}
+		}
+		return 0;
+	}
+
 	if (!req->all)
 		return npc_delete_flow(rvu, req->entry, pcifunc);
 
-- 
2.31.1

